File: packages/couriers/src/steadfast.ts
// Steadfast courier adapter. PURE: fetch + creds injected per-call.
//
// Base https://portal.steadfast.com.bd/api/v1. Auth via Api-Key / Secret-Key
// headers. No sandbox exists → this is contract-tested against the documented
// request/response shapes; live verification is deferred until a merchant
// account exists (see Phase-1 brief).
//
//   create  POST /create_order  {invoice, recipient_name, recipient_phone,
//                                recipient_address, cod_amount, note}
//             → consignment.{consignment_id, tracking_code}
//   status  GET  /status_by_cid/{id}  → {delivery_status}
//   balance GET  /get_balance         → {current_balance}
import type {
  CourierAdapter,
  CourierCreds,
  FetchLike,
  ConsignmentInput,
  ConsignmentResult,
  StatusResult,
} from "./types";
import { mapSteadfastStatus } from "./statusMap";

const BASE_URL = "https://portal.steadfast.com.bd/api/v1";

export interface SteadfastProviderOptions {
  fetch: FetchLike;
}

export class SteadfastProvider implements CourierAdapter {
  readonly provider = "steadfast" as const;

  private readonly fetch: FetchLike;

  constructor(opts: SteadfastProviderOptions) {
    this.fetch = opts.fetch;
  }

  private headers(creds: CourierCreds): Record<string, string> {
    return {
      "Api-Key": creds.apiKey,
      "Secret-Key": creds.secretKey,
      "Content-Type": "application/json",
    };
  }

  async createConsignment(input: ConsignmentInput, creds: CourierCreds): Promise<ConsignmentResult> {
    const res = await this.fetch(`${BASE_URL}/create_order`, {
      method: "POST",
      headers: this.headers(creds),
      body: JSON.stringify({
        invoice: input.invoice,
        recipient_name: input.recipient_name,
        recipient_phone: input.recipient_phone,
        recipient_address: input.recipient_address,
        cod_amount: Math.round(input.cod_amount), // Steadfast expects an integer
        note: input.note ?? "",
      }),
    });

    const body = (await res.json()) as {
      status?: number;
      consignment?: { consignment_id?: number | string; tracking_code?: string };
      message?: string;
    };

    const consignment = body.consignment;
    if (!consignment?.consignment_id || !consignment.tracking_code) {
      throw new Error(`Steadfast create_order failed: ${body.message ?? res.status}`);
    }

    return {
      consignmentId: String(consignment.consignment_id),
      trackingCode: consignment.tracking_code,
      raw: body,
    };
  }

  async getStatus(consignmentId: string, creds: CourierCreds): Promise<StatusResult> {
    const res = await this.fetch(`${BASE_URL}/status_by_cid/${consignmentId}`, {
      method: "GET",
      headers: this.headers(creds),
    });

    const body = (await res.json()) as { delivery_status?: string; status?: number };
    const mapped = mapSteadfastStatus(body.delivery_status ?? "");

    return {
      status: mapped.shipment_status,
      fulfillment: mapped.order_fulfillment_status,
      raw: body,
    };
  }

  async getBalance(creds: CourierCreds): Promise<number> {
    const res = await this.fetch(`${BASE_URL}/get_balance`, {
      method: "GET",
      headers: this.headers(creds),
    });

    const body = (await res.json()) as { current_balance?: number; status?: number };
    return body.current_balance ?? 0;
  }
}

File: apps/api/app/couriers/steadfast.py
"""Steadfast adapter (httpx). Port of packages/couriers/src/steadfast.ts.

PURE-ish: takes an injected httpx.AsyncClient + per-call creds (never holds
secrets). Base https://portal.steadfast.com.bd/api/v1, auth via Api-Key /
Secret-Key headers. No sandbox exists — contract-shaped; live verification is
deferred until a merchant account exists (same as the TS side).

The sweep depends on the `StatusFetcher` Protocol, so tests inject a fake with
no network.
"""
from dataclasses import dataclass
from typing import Any, Protocol

import httpx

from .status_map import map_steadfast_status


@dataclass
class StatusResult:
    shipment_status: str
    fulfillment: str
    raw: dict[str, Any]


class StatusFetcher(Protocol):
    async def get_status(self, consignment_id: str, creds: dict[str, str]) -> StatusResult: ...


class SteadfastClient:
    def __init__(self, http: httpx.AsyncClient, base_url: str) -> None:
        self._http = http
        self._base = base_url.rstrip("/")

    @staticmethod
    def _headers(creds: dict[str, str]) -> dict[str, str]:
        api_key = creds.get("apiKey")
        secret_key = creds.get("secretKey")
        if not api_key or not secret_key:
            raise ValueError("Steadfast credentials incomplete (apiKey/secretKey required)")
        return {"Api-Key": api_key, "Secret-Key": secret_key, "Content-Type": "application/json"}

    async def get_status(self, consignment_id: str, creds: dict[str, str]) -> StatusResult:
        res = await self._http.get(
            f"{self._base}/status_by_cid/{consignment_id}", headers=self._headers(creds)
        )
        body: dict[str, Any] = res.json()
        mapped = map_steadfast_status(body.get("delivery_status") or "")
        return StatusResult(
            shipment_status=mapped["shipment_status"],
            fulfillment=mapped["order_fulfillment_status"],
            raw=body,
        )
